src/python/turicreate/_deps/__init__.py

```diff
@@ -9,59 +9,346 @@
 from distutils.version import StrictVersion as _StrictVersion
 import logging as _logging
 import re as _re
+import sys
+import six
+from six.moves import reload_module
+from types import ModuleType
 
+if six.PY2:
+    import imp
+else:
+    import importlib
+
+def run_from_ipython():
+    # Taken from https://stackoverflow.com/questions/5376837
+    try:
+        __IPYTHON__
+        return True
+    except NameError:
+        return False
+
+## builtin __import__(path, globals, locals, fromlist)
+# referenced from https://github.com/mnmelo/lazy_import/blob/master/lazy_import/__init__.py
+if six.PY2:
+    import imp
+    # reentrant lock
+    # https://docs.python.org/2/library/imp.html#imp.acquire_lock
+    class _ImportLockContext:
+        def __enter__(self):
+            imp.acquire_lock()
+        def __exit__(self, exc_type, exc_value, exc_traceback):
+            imp.release_lock()
+else:
+    # importlib uses _imp internally; check source you will know ;-)
+    from importlib._bootstrap import _ImportLockContext
+
+# should always be guared by _ImportLockContext
+def default_init_func(name, *args, **kwargs):
+    """
+    called within _ImportLockContext(); no import lock should be hold inside
+    """
+    _ret = sys.modules.get(name, None)
+    if _ret is not None:
+        assert name not in sys.modules.keys(), ("sys.modules[%s] cannot be None"
+                                                " during module loading" % name)
+    if six.PY2:
+        # imp importing submodule is too tricky
+        # moulde and pkg naming should follow pep8
+        if not all([x.isalnum() or x in '._' for x in name]):
+            raise ValueError('invalid module name')
+        exec("import %s as _py2_ret" % name)
+        return _py2_ret
+    else:
+        return importlib.import_module(name)
+
+class LazyModuleLoader(ModuleType, object):
+    """ defer to load a module. If it's loaded, no exception but warning will be given. """
```

Comment:
Change this comment?  I believe this is now just a no-op?

Reply:
oh yeah, I forget to clean it up. Will do.